spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java

```diff
@@ -246,11 +270,202 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
+				taskExecutionInformation.getTaskDeploymentProperties().put(dataflowAccessTokenPropertyKey, token);
+			}
+		}
+	}
+
+	/**
+	 * Stores the platform specific execution id for a given task execution
+	 *
+	 * @param taskExecution task execution id to associate the external execution id with
+	 * @param taskDeploymentId platform specific execution id
+	 */
+	private void saveExternalExecutionId(TaskExecution taskExecution, String taskDeploymentId) {
+		if (!StringUtils.hasText(taskDeploymentId)) {
+			throw new IllegalStateException("Deployment ID is null for the task:" + taskExecution.getTaskName());
+		}
+		this.updateExternalExecutionId(taskExecution.getExecutionId(), taskDeploymentId);
+	}
+
+	/**
+	 * Updates the deployment properties on the provided {@code AppDeploymentRequest}
+	 *
+	 * @param commandLineArgs command line args for the task execution
+	 * @param platformName name of the platform configuration to use
+	 * @param taskExecutionInformation details about the task execution request
+	 * @param taskExecution task execution data
+	 * @param previousManifest manifest from the last execution of the same task definition
+	 * @return an updated {@code AppDeploymentRequest}
+	 */
+	private AppDeploymentRequest updateDeploymentProperties(List<String> commandLineArgs, String platformName, TaskExecutionInformation taskExecutionInformation, TaskExecution taskExecution, TaskManifest previousManifest) {
+		AppDeploymentRequest appDeploymentRequest;
+		TaskExecutionInformation info = new TaskExecutionInformation();
+		info.setTaskDefinition(taskExecutionInformation.getTaskDefinition());
+		info.setAppResource(taskExecutionInformation.getAppResource());
+		info.setComposed(taskExecutionInformation.isComposed());
+		info.setMetadataResource(taskExecutionInformation.getMetadataResource());
+		info.setOriginalTaskDefinition(taskExecutionInformation.getOriginalTaskDefinition());
+		info.setTaskDeploymentProperties(previousManifest.getTaskDeploymentRequest().getDeploymentProperties());
+
+		appDeploymentRequest = this.taskAppDeploymentRequestCreator.
+				createRequest(taskExecution, info, commandLineArgs, platformName);
+		return appDeploymentRequest;
+	}
+
+	/**
+	 * A task should not be allowed ot be launched when an upgrade is required and one is running (allowing the upgrade
+	 * to proceed may kill running task instances of that definition on certain platforms).
+	 *
+	 * @param taskName task name to check or lock
+	 * @param platformName the platform configuration to confirm if the task is being run on
+	 */
+	private void validateAndLockUpgrade(String taskName, String platformName, TaskExecution taskExecution) {
+		Page<TaskExecution> runningTaskExecutions =
+				this.taskExplorer.findRunningTaskExecutions(taskName, PageRequest.of(0, 1));
+
+		//TODO add force flag to allow overriding this
+		if(!(runningTaskExecutions.getTotalElements() == 1 && runningTaskExecutions.toList().get(0).getExecutionId() == taskExecution.getExecutionId()) &&
```

Comment:
Actually, that should be an `||` instead of an `&&` on that first if...I'll fix.

Reply:
If the task definition is already running, we do not allow a user to do an upgrade because the destroy would kill the currently running task (on cf).  This logic handles that check.  If it is not running, we store the name of the task definition and the platform it is running on in a Map to prevent others from upgrading it while this one is.